sapling_key_components.py

```diff
@@ -81,7 +80,59 @@ def default_d(self):
     @cached
     def default_pkd(self):
         return group_hash(b'Zcash_gd', self.default_d()) * self.ivk()
-
+    
+    @cached
+    def expanded_spending_key(self):
+        return ExpandedSpendingKey(self.ask(), self.nsk(), self.ovk())
+        
+    @cached
+    def full_viewing_key(self):
+        return self.expanded_spending_key().full_viewing_key()
+            
+    @cached 
+    def default_address(self):
+        return self.full_viewing_key().incoming_viewing_key().address(self.default_d())
+    
+class ExpandedSpendingKey(object):
+    def __init__(self, ask, nsk, ovk):
```

Comment:
I would define a mix-in class that makes ``ak()``, ``nk()``, and ``ivk()`` also available on ``ExpandedSpendingKey``, like this:
```
# This assumes subclasses define ask(), nsk(), and ovk().
class SpendingKeyMixin(object):
    @cached
    def ak(self):
        return SPENDING_KEY_BASE * self.ask()
 
    @cached
    def nk(self):
        return PROVING_KEY_BASE * self.nsk()
 
    @cached
    def ivk(self):
        return Fr.from_bytes(crh_ivk(bytes(self.ak()), bytes(self.nk())))

    def full_viewing_key(self):
        return FullViewingKey(self.ak(), self.nk(), self.ovk())


class SpendingKey(SpendingKeyMixin):
    def __init__(self, data):
        self.data = data
 
    @cached
    def ask(self):
        return Fr.from_bytes(prf_expand(self.data, b'\0'))
 
    @cached
    def nsk(self):
        return Fr.from_bytes(prf_expand(self.data, b'\1'))
 
    @cached
    def ovk(self):
        return prf_expand(self.data, b'\2')[:32]

    @cached
    def default_d(self):
        i = 0
        while True:
            d = prf_expand(self.data, bytes([3, i]))[:11]
            if group_hash(b'Zcash_gd', d):
                return d
            i += 1
            assert i < 256
 
    @cached
    def default_pkd(self):
        return group_hash(b'Zcash_gd', self.default_d()) * self.ivk()

    def expanded_spending_key(self):
        return ExpandedSpendingKey(self.ask(), self.nsk(), self.ovk())


class ExpandedSpendingKey(SpendingKeyMixin):
    def __init__(self, ask, nsk, ovk):
        self._ask = ask 
        self._nsk = nsk
        self._ovk = ovk

    def ask(self):
        return self._ask

    def nsk(self):
        return self._nsk

    def ovk(self):
        return self._ovk
```